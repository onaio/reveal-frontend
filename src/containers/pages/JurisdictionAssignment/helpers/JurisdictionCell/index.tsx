/** component in the jurisdictionTable selector that will render a node according to a few rules:
 * 1) as a link if the node has children
 * 2) as plain text if the node does not have children.
 */
import React from 'react';
import { Link } from 'react-router-dom';
import { TreeNode } from '../../../../../store/ducks/opensrp/hierarchies/types';

/**
 * Props for NodeCell
 */
export interface NodeCellProps {
  node: TreeNode /** the current jurisdiction */;
  baseUrl: string;
}

/**
 * NodeCell
 *
 * Displays the jurisdiction.  If the jurisdiction is not a leaf node, then it is
 * displayed as a link, so that one can traverse down the tree.
 * @param props - the props
 */
const NodeCell = (props: NodeCellProps) => {
  const { node, baseUrl } = props;

  // isLeafNode if node does not have children
  const isLeafNode = !node.hasChildren();

  if (isLeafNode) {
    return <span key={`${node.model.id}-span`}>{node.model.label}</span>;
  }

  return (
    <Link to={`${baseUrl}/${node.model.id}`} key={`${node.model.id}-span`}>
      {node.model.label}
    </Link>
  );
};

export { NodeCell };
