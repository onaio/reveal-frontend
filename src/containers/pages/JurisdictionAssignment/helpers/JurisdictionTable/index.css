/* allows tooltip to show on disabled button */
.btn[disabled] {
  pointer-events: none;
}
