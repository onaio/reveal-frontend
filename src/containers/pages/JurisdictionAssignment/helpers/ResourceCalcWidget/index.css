/* Chrome, Safari, Edge, Opera */

#resource-calc-widget input::-webkit-outer-spin-button,
#resource-calc-widget input::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

/* Firefox */
#resource-calc-widget input[type='number'] {
  -moz-appearance: textfield;
}

#resource-calc-widget input {
  width: 45px;
}
