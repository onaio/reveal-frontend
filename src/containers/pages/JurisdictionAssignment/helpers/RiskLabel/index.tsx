/** Rendered as a table cell and shows the risk label for a given jurisdiction node */
import React from 'react';
import { TreeNode } from '../../../../../store/ducks/opensrp/hierarchies/types';
import { JurisdictionsMetadata } from '../../../../../store/ducks/opensrp/jurisdictionsMetadata';

export interface RiskLabelProps {
  metadata: JurisdictionsMetadata[];
  node: TreeNode;
}

/** finds the nodes risk label from the fetched metadata and displays it */
export const RiskLabel = (props: RiskLabelProps) => {
  const { metadata, node } = props;

  // could get more that one meta objects
  const metasOfInterest = metadata.filter(meta => meta.key === node.model.id);

  // pick the first one with a value
  const metaOfInterest = metasOfInterest.filter(meta => meta.value !== undefined);
  const theSingleMeta = metaOfInterest[0];

  if (theSingleMeta && theSingleMeta.value) {
    return <>{theSingleMeta.value}</>;
  }
  return <>{'--'}</>;
};
