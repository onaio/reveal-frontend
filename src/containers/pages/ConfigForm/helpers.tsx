import React from 'react';
import Loading from '../../../components/page/Loading';
import { NoDataComponent } from '../../../components/Table/NoDataComponent';
import { OPENSRP_API_BASE_URL } from '../../../configs/env';
import { getAcessTokenOrRedirect, growl } from '../../../helpers/utils';
import { newGetPayloadOptions } from '../../../services/opensrp';

/** drill down props */
export const drillDownProps = {
  paginate: false,
  renderNullDataComponent: () => <NoDataComponent />,
};

export const defaultConfigProps = {
  LoadingComponent: <Loading />,
  accessToken: getAcessTokenOrRedirect,
  baseURL: OPENSRP_API_BASE_URL,
  customAlert: growl,
  getPayload: newGetPayloadOptions,
};
