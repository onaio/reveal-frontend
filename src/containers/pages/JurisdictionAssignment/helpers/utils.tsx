import { useState } from 'react';
import { displayError } from '../../../../helpers/errors';
import { TreeNode } from '../../../../store/ducks/opensrp/hierarchies/types';
import { nodeIsSelected } from '../../../../store/ducks/opensrp/hierarchies/utils';

/** tells whether we should select the parent checkbox in the table header
 * @param currentParentNode - the current parent node
 * @param currentChildren - current children
 */
export const checkParentCheckbox = (
  currentParentNode: TreeNode | undefined,
  currentChildren: TreeNode[]
) => {
  let selected = true;
  if (currentParentNode) {
    selected = selected && nodeIsSelected(currentParentNode);
  } else {
    // when currentParent node is undefined but there are currentChildren usually at the top level
    if (currentChildren.length > 0) {
      currentChildren.forEach(node => {
        selected = selected && nodeIsSelected(node);
      });
    } else {
      selected = false;
    }
  }
  return selected;
};

export const useHandleBrokenPage = () => {
  const [broken, setBroken] = useState<boolean>(false);
  const [errorMessage, setErrorMessage] = useState<string | undefined>();

  /** Convenience function to handle cases where we must abort and tell the user we have done so */
  const handleBrokenPage = (message: string) => {
    displayError(Error(message));
    setErrorMessage(message);
    setBroken(true);
  };

  return { broken, errorMessage, handleBrokenPage };
};
