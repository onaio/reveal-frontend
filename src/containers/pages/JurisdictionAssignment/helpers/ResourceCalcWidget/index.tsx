/** Resource Calculation component
 * going to be a side panel that calculates the ETA for an available
 * number of teams to be visit a variable number of structures in selected
 * jurisdictions.
 */

import reducerRegistry from '@onaio/redux-reducer-registry';
import { Field, Form, Formik } from 'formik';
import React from 'react';
import { connect } from 'react-redux';
import { Store } from 'redux';
import { format } from 'util';
import {
  AT_A_RATE_OF,
  NO_JURISDICTION_SELECTIONS_FOUND,
  RESOURCE_ESTIMATE_FOR,
  STRUCTURES_PER_TEAM_PER_DAY,
  TEAMS,
} from '../../../../../configs/lang';
import hierarchyReducer, {
  getCurrentParentNode,
  getParentNodeInSelectedTree,
  reducerName as hierarchyReducerName,
} from '../../../../../store/ducks/opensrp/hierarchies';
import { TreeNode } from '../../../../../store/ducks/opensrp/hierarchies/types';
import { getNodeStructureCount } from '../../../../../store/ducks/opensrp/hierarchies/utils';
import './index.css';

reducerRegistry.register(hierarchyReducerName, hierarchyReducer);

/** props */
export interface ResourceCalculationProps {
  rootId: string;
  planId: string;
  currentParentNode?: TreeNode;
  selectedTreeCurrentParentNode?: TreeNode;
  currentParentId?: string;
}

/** helper function that computes the estimate time to completion given a
 * the total structures and rate of completion per team
 * @param totalStructures - total structures in these jurisdiction
 * @param structuresPerTeam - rate of completion; how many structures a team can handler
 * @param teams - tne number of teams
 */
export const computeEstimate = (
  totalStructures: number,
  structuresPerTeam: number,
  teams: number
) => {
  if (!teams || !structuresPerTeam) {
    return 0;
  }
  return Math.ceil(totalStructures / (teams * structuresPerTeam));
};

/** The component that renders the resource calculation info */
const ResourceCalculation = (props: ResourceCalculationProps) => {
  const { selectedTreeCurrentParentNode, currentParentNode } = props;

  if (!selectedTreeCurrentParentNode) {
    const parentNodeLabel = currentParentNode ? `${currentParentNode.model.label}: ` : '';
    return (
      <div id="resource-calc-widget">
        {`${parentNodeLabel} ${NO_JURISDICTION_SELECTIONS_FOUND}`}
      </div>
    );
  }
  const jurisdictionName = selectedTreeCurrentParentNode.model.label;
  const structuresCount = getNodeStructureCount(selectedTreeCurrentParentNode);

  return (
    <div id="resource-calc-widget">
      <h3 className="section-title">{format(RESOURCE_ESTIMATE_FOR, jurisdictionName)}</h3>

      <Formik
        initialValues={{ structuresCount: 0, teamsCount: 0 }}
        // tslint:disable-next-line: jsx-no-lambda
        onSubmit={() => {
          return;
        }}
      >
        {formikProps => {
          const { structuresCount: formikStructureSCount, teamsCount } = formikProps.values;
          const numDays = computeEstimate(structuresCount, formikStructureSCount, teamsCount);
          return (
            <Form>
              <p>
                {`${numDays} days`}
                <br />
                {AT_A_RATE_OF} <Field type="number" name="structuresCount" />
                {'  '}
                {STRUCTURES_PER_TEAM_PER_DAY} <Field type="number" name="teamsCount" /> {TEAMS}
              </p>
            </Form>
          );
        }}
      </Formik>
    </div>
  );
};

export { ResourceCalculation };

type MapStateToProps = Pick<
  ResourceCalculationProps,
  'currentParentNode' | 'selectedTreeCurrentParentNode'
>;

export const selectedTreeParentNodeSelector = getParentNodeInSelectedTree();
export const parentNodeSelector = getCurrentParentNode();

/** map props to store state selectors
 * @param state - the store
 * @param ownProps - component props
 */
const mapStateToProps = (
  state: Partial<Store>,
  ownProps: ResourceCalculationProps
): MapStateToProps => {
  const filters = {
    currentParentId: ownProps.currentParentId,
    planId: ownProps.planId,
    rootJurisdictionId: ownProps.rootId,
  };
  const currentParentNode = parentNodeSelector(state, filters);
  const selectedTreeCurrentParentNode = selectedTreeParentNodeSelector(state, filters);
  return { currentParentNode, selectedTreeCurrentParentNode };
};

const ConnectedResourceWidget = connect(mapStateToProps)(ResourceCalculation);
export { ConnectedResourceWidget };
