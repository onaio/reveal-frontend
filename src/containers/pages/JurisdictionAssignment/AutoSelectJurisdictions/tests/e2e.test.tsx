import reducerRegistry from '@onaio/redux-reducer-registry';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { act } from 'react-dom/test-utils';
import { Provider } from 'react-redux';
import { MemoryRouter, Route, Switch } from 'react-router';
import ConnectedAutoSelectView from '..';
import { AUTO_ASSIGN_JURISDICTIONS_URL } from '../../../../../constants';
import { renderTable } from '../../../../../helpers/testUtils';
import store from '../../../../../store';
import hierarchiesReducer, {
  deforest,
  deselectAllNodes,
  reducerName as hierarchiesReducerName,
} from '../../../../../store/ducks/opensrp/hierarchies';
import { zambiaHierarchy } from '../../../../../store/ducks/opensrp/hierarchies/tests/fixtures';
import plansReducer, { reducerName } from '../../../../../store/ducks/opensrp/PlanDefinition';
import { plans } from '../../../../../store/ducks/opensrp/PlanDefinition/tests/fixtures';
import { ResourceCalculation } from '../../helpers/ResourceCalcWidget';
import { autoSelectionJurisdictionMeta } from './fixtures';

reducerRegistry.register(reducerName, plansReducer);
reducerRegistry.register(hierarchiesReducerName, hierarchiesReducer);

/* tslint:disable-next-line no-var-requires */
const fetch = require('jest-fetch-mock');

jest.mock('reactstrap', () => {
  const original = require.requireActual('reactstrap');
  return { ...original, Tooltip: () => null };
});

jest.mock('../../../AssigmentMapWrapper', () => {
  const mockComponent2 = (_: any) => <div id="mockComponent2">Assignment Wrapper</div>;
  return {
    ConnectedAssignmentMapWrapper: mockComponent2,
  };
});

describe('e2e test auto-selection flow', () => {
  afterEach(() => {
    store.dispatch(deselectAllNodes('22bc44dd-752d-4c20-8761-617361b4f1e7'));
    store.dispatch(deforest());
  });
  it('from slider auto-selection to refine jurisdictions', async () => {
    jest.setTimeout(30000);
    const envModule = require('../../../../../configs/env');
    envModule.ASSIGNMENT_PAGE_SHOW_MAP = true;

    const App = () => {
      return (
        <Switch>
          <Route
            exact={true}
            path={`${AUTO_ASSIGN_JURISDICTIONS_URL}/:planId/:rootId/:parentId`}
            component={ConnectedAutoSelectView}
          />
          <Route
            exact={true}
            path={`${AUTO_ASSIGN_JURISDICTIONS_URL}/:planId/:rootId`}
            component={ConnectedAutoSelectView}
          />
        </Switch>
      );
    };

    const plan = plans[0];
    const rootId = '22bc44dd-752d-4c20-8761-617361b4f1e7';
    fetch
      .once(JSON.stringify(autoSelectionJurisdictionMeta), { status: 200 })
      .once(JSON.stringify(plan), { status: 200 })
      .once(JSON.stringify(zambiaHierarchy), { status: 200 });

    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[
            {
              hash: '',
              pathname: `${AUTO_ASSIGN_JURISDICTIONS_URL}/${plan.identifier}/${rootId}`,
              search: '',
              state: {},
            },
          ]}
        >
          <App />
        </MemoryRouter>
      </Provider>
    );

    await act(async () => {
      await new Promise(resolve => setImmediate(resolve));
      wrapper.update();
    });

    expect(wrapper.text()).toMatchSnapshot('general page snapshot');

    /** we start with the slider component where we have tailored the metaData
     * to facilitate 2 nodes autoSelection:
     */
    // CWY_25 (178 structures) is selected by default during mount since it has a threshold of 80 > 0
    expect(wrapper.find('div.slider-section').text()).toMatchInlineSnapshot(`"Risk  0%00100"`);
    expect(wrapper.find('div.info-section').text()).toMatchInlineSnapshot(
      `"NUMBER OF STRUCTURES IN SELECTED JURISDICTIONS178"`
    );

    // click next to proceed to the structureSummary view
    wrapper.find('JurisdictionSelectionsSlider button.btn-success').simulate('click');

    act(() => {
      wrapper.update();
    });

    expect(wrapper.text()).toMatchSnapshot('structure summary step snapshot');

    /** structure summary step, check that drill-down only has the correct nodes */

    renderTable(wrapper, 'should have single node(parent node)');
    const tbodyRow = wrapper.find('tbody tr');
    expect(tbodyRow.length).toEqual(1);
    expect(
      tbodyRow
        .at(0)
        .text()
        .includes('Zambia')
    ).toBeTruthy();
    // expect only one tbody row
    expect(wrapper.find('tbody tr').length).toEqual(1);

    expect(
      wrapper
        .find('tbody tr')
        .at(0)
        .text()
    ).toMatchInlineSnapshot(`"Zambia178"`);

    // test drilldown
    tbodyRow
      .at(0)
      .find('NodeCell Link a')
      .simulate('click', { button: 0 });

    wrapper.update();
    renderTable(wrapper, 'after first click');

    // expect only one tbody row
    expect(wrapper.find('tbody tr').length).toEqual(1);
    expect(
      wrapper
        .find('tbody tr')
        .at(0)
        .text()
    ).toMatchInlineSnapshot(`"Eastern178"`);

    wrapper
      .find('tbody tr')
      .at(0)
      .find('NodeCell a')
      .simulate('click', { button: 0 });

    wrapper.update();
    renderTable(wrapper, 'after second click');

    // expect only one tbody row
    expect(wrapper.find('tbody tr').length).toEqual(1);
    expect(
      wrapper
        .find('tbody tr')
        .at(0)
        .text()
    ).toMatchInlineSnapshot(`"Sinda178"`);

    wrapper
      .find('tbody tr')
      .at(0)
      .find('NodeCell a')
      .simulate('click', { button: 0 });

    wrapper.update();
    renderTable(wrapper, 'after third click');

    // expect only one tbody row
    expect(wrapper.find('tbody tr').length).toEqual(1);
    expect(
      wrapper
        .find('tbody tr')
        .at(0)
        .text()
    ).toMatchInlineSnapshot(`"Chiwuyu178"`);

    wrapper
      .find('tbody tr')
      .at(0)
      .find('NodeCell a')
      .simulate('click', { button: 0 });

    wrapper.update();
    renderTable(wrapper, 'after third click');

    // expect only one tbody row
    expect(wrapper.find('tbody tr').length).toEqual(1);
    expect(
      wrapper
        .find('tbody tr')
        .at(0)
        .text()
    ).toMatchInlineSnapshot(`"CWY_25178"`);

    // CWY_25 is the last child it should not be clickable
    expect(toJson(wrapper.find('tbody tr NodeCell span'))).toMatchSnapshot(
      'should not have btn-link or onClick handler'
    );

    /** try resource calculation for CWY_25 */

    // simulate structure count change so that we can have 10 structures per team
    wrapper
      .find('input[name="structuresCount"]')
      .simulate('change', { target: { value: 10, name: 'structuresCount' } });
    wrapper.update();

    wrapper
      .find('input[name="teamsCount"]')
      .simulate('change', { target: { value: 3, name: 'teamsCount' } });
    wrapper.update();

    // see if there has been any change to the number of days, should start with 6 days
    // 178 structures / 30 structures per team for 3 teams => 178 / 30
    expect(wrapper.find('form p').text()).toMatchInlineSnapshot(
      `"6 daysat a rate of   structures per team per day with  Teams"`
    );

    // click next goto the refine jurisdiction selections section
    wrapper.find('SelectedStructuresTable button.btn-success').simulate('click');

    await act(async () => {
      await new Promise(resolve => setImmediate(resolve));
      wrapper.update();
    });

    expect(wrapper.text()).toMatchSnapshot('refine selected jurisdictions component');

    /** we should now be in the jurisdictions refinement component,  */

    // here we check that the whole tree is shown i.e. not just nodes in
    // a path that has selected nodes.
    // there is now more than one node(the ones that is selected)
    expect(wrapper.find('tbody tr').length).toEqual(23);

    // drill up to parent node. // Zambia: it should now have more children
    // after drill-down than those showed in the structure summary step

    expect(
      wrapper
        .find('JurisdictionTable HeaderBreadcrumb BreadcrumbItem')
        .at(1)
        .text()
    ).toMatchInlineSnapshot(`"Zambia"`);
    wrapper
      .find('JurisdictionTable HeaderBreadcrumb BreadcrumbItem')
      .at(1)
      .find('a')
      .simulate('click', { button: 0 });

    await act(async () => {
      await new Promise(resolve => setImmediate(resolve));
      wrapper.update();
    });

    // all children of Zambia are shown: only one of them is in a path that has
    // selected nodes
    expect(wrapper.find('tbody tr').length).toEqual(6);

    // simulate a click to check one of the immediate children of Zambia: Western
    // => effectively checking all leaf nodes that are descendants of Western
    wrapper.find('tbody tr').forEach(tr => {
      if (tr.text().includes('Western')) {
        // check the checkbox
        tr.find('input').simulate('change', { target: { name: '', checked: true } });
      }
    });

    await act(async () => {
      await new Promise(resolve => setImmediate(resolve));
      wrapper.update();
    });

    // check for new selections
    wrapper.find('tbody tr').forEach(tr => {
      if (tr.text().includes('Western')) {
        expect(
          tr
            .find('td input')
            .at(0)
            .props().checked
        ).toBeTruthy();
      } else {
        expect(
          tr
            .find('td input')
            .at(0)
            .props().checked
        ).toBeFalsy();
      }
    });
  });

  it('when there are no selected jurisdictions(without metaData)', async () => {
    jest.setTimeout(20000);
    const envModule = require('../../../../../configs/env');
    envModule.ASSIGNMENT_PAGE_SHOW_MAP = true;

    const App = () => {
      return (
        <Switch>
          <Route
            exact={true}
            path={`${AUTO_ASSIGN_JURISDICTIONS_URL}/:planId/:rootId/:parentId`}
            component={ConnectedAutoSelectView}
          />
          <Route
            exact={true}
            path={`${AUTO_ASSIGN_JURISDICTIONS_URL}/:planId/:rootId`}
            component={ConnectedAutoSelectView}
          />
        </Switch>
      );
    };

    const plan = plans[0];
    const rootId = '22bc44dd-752d-4c20-8761-617361b4f1e7';
    fetch
      .once(JSON.stringify([]), { status: 200 })
      .once(JSON.stringify(plan), { status: 200 })
      .once(JSON.stringify(zambiaHierarchy), { status: 200 });

    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter
          initialEntries={[
            {
              hash: '',
              pathname: `${AUTO_ASSIGN_JURISDICTIONS_URL}/${plan.identifier}/${rootId}`,
              search: '',
              state: {},
            },
          ]}
        >
          <App />
        </MemoryRouter>
      </Provider>
    );

    await act(async () => {
      await new Promise(resolve => setImmediate(resolve));
      wrapper.update();
    });

    expect(wrapper.text()).toMatchSnapshot('general page snapshot');

    /** we start with the slider component : there is no metadata so no matter what
     * threshold we set there will be no selected jurisdictions
     */
    // CWY_25 (178 structures) is selected by default during mount since it has a threshold of 80 > 0
    expect(wrapper.find('div.slider-section').text()).toMatchInlineSnapshot(`"Risk  0%00100"`);
    expect(wrapper.find('div.info-section').text()).toMatchInlineSnapshot(
      `"NUMBER OF STRUCTURES IN SELECTED JURISDICTIONS0"`
    );

    // click next to proceed to the structureSummary view
    wrapper.find('JurisdictionSelectionsSlider button.btn-success').simulate('click');

    act(() => {
      wrapper.update();
    });

    expect(wrapper.text()).toMatchSnapshot('structure summary step snapshot');

    /** structure summary step, this would not have any records */

    renderTable(wrapper, 'will not have any nodes');
    const tbodyRow = wrapper.find('tbody tr');
    expect(tbodyRow.length).toEqual(0);
    expect(wrapper.find('SelectedStructuresTable').text()).toMatchInlineSnapshot(
      `"....NameStructures CountThere are no descendant jurisdictions that are targeted.Continue to next step"`
    );

    /** What of the resource calculation component */

    // is an error page => will refactor this to an info message instead
    expect(wrapper.find(ResourceCalculation).text()).toMatchInlineSnapshot(
      `"Zambia:  There are no descendant jurisdictions that are targeted."`
    );

    // click next goto the refine jurisdiction selections section
    wrapper.find('SelectedStructuresTable button.btn-success').simulate('click');

    await act(async () => {
      await new Promise(resolve => setImmediate(resolve));
      wrapper.update();
    });

    // expect(wrapper.text()).toMatchSnapshot('refine selected jurisdictions component');

    /** we should now be in the jurisdictions refinement component,  */

    // here we check that the whole tree is shown
    expect(wrapper.text()).toMatchSnapshot('full page rendered text');

    const bodyRows = wrapper.find('tbody tr');
    // test drilldown
    bodyRows
      .at(0)
      .find('NodeCell Link a')
      .simulate('click', { button: 0 });

    await act(async () => {
      await new Promise(resolve => setImmediate(resolve));
      wrapper.update();
    });

    // all children of Zambia are shown:
    expect(wrapper.find('tbody tr').length).toEqual(6);

    // simulate a click to check one of the immediate children of Zambia: Western
    // => effectively checking all leaf nodes that are descendants of Western
    wrapper.find('tbody tr').forEach(tr => {
      if (tr.text().includes('Western')) {
        // check the checkbox
        tr.find('input').simulate('change', { target: { name: '', checked: true } });
      }
    });

    await act(async () => {
      await new Promise(resolve => setImmediate(resolve));
      wrapper.update();
    });

    // check for new selections
    wrapper.find('tbody tr').forEach(tr => {
      if (tr.text().includes('Western')) {
        expect(
          tr
            .find('td input')
            .at(0)
            .props().checked
        ).toBeTruthy();
      } else {
        expect(
          tr
            .find('td input')
            .at(0)
            .props().checked
        ).toBeFalsy();
      }
    });

    /** the resource calculation component updates accordingly */
    expect(wrapper.find(ResourceCalculation).text()).toMatchInlineSnapshot(
      `"Resource Estimate for Zambia0 daysat a rate of   structures per team per day with  Teams"`
    );

    // check resource calc behavior when drilling down to children node that does not
    // have selected jurisdictions
    wrapper.find('tbody tr').forEach(tr => {
      if (tr.text().includes('Eastern')) {
        // check the checkbox
        tr.find('NodeCell Link a').simulate('click', { button: 0 });
      }
    });
    await act(async () => {
      await new Promise(resolve => setImmediate(resolve));
      wrapper.update();
    });

    /** the resource calculation component updates accordingly */
    expect(wrapper.find(ResourceCalculation).text()).toMatchInlineSnapshot(
      `"Eastern:  There are no descendant jurisdictions that are targeted."`
    );
  });
});
