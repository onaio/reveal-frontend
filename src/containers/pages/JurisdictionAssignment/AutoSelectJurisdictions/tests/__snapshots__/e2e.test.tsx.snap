// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`e2e test auto-selection flow from slider auto-selection to refine jurisdictions: after first click 1`] = `"NameStructures Count"`;

exports[`e2e test auto-selection flow from slider auto-selection to refine jurisdictions: after first click 2`] = `"Eastern178"`;

exports[`e2e test auto-selection flow from slider auto-selection to refine jurisdictions: after first click 3`] = `"Total178"`;

exports[`e2e test auto-selection flow from slider auto-selection to refine jurisdictions: after second click 1`] = `"NameStructures Count"`;

exports[`e2e test auto-selection flow from slider auto-selection to refine jurisdictions: after second click 2`] = `"Sinda178"`;

exports[`e2e test auto-selection flow from slider auto-selection to refine jurisdictions: after second click 3`] = `"Total178"`;

exports[`e2e test auto-selection flow from slider auto-selection to refine jurisdictions: after third click 1`] = `"NameStructures Count"`;

exports[`e2e test auto-selection flow from slider auto-selection to refine jurisdictions: after third click 2`] = `"Chiwuyu178"`;

exports[`e2e test auto-selection flow from slider auto-selection to refine jurisdictions: after third click 3`] = `"Total178"`;

exports[`e2e test auto-selection flow from slider auto-selection to refine jurisdictions: after third click 4`] = `"NameStructures Count"`;

exports[`e2e test auto-selection flow from slider auto-selection to refine jurisdictions: after third click 5`] = `"CWY_25178"`;

exports[`e2e test auto-selection flow from slider auto-selection to refine jurisdictions: after third click 6`] = `"Total178"`;

exports[`e2e test auto-selection flow from slider auto-selection to refine jurisdictions: general page snapshot 1`] = `"Planning toolA2-Lusaka Akros Test Focus 2A2-Lusaka Akros Test Focus 2Auto target jurisdictions by riskRefine selected jurisdictionsAdjust slider to auto-target jurisdictionsRisk  0%00100NUMBER OF STRUCTURES IN SELECTED JURISDICTIONS178Continue to next step"`;

exports[`e2e test auto-selection flow from slider auto-selection to refine jurisdictions: refine selected jurisdictions component 1`] = `"Planning toolA2-Lusaka Akros Test Focus 2A2-Lusaka Akros Test Focus 2Auto target jurisdictions by riskRefine selected jurisdictionsAssignment WrapperResource Estimate for Chiwuyu6 daysat a rate of   structures per team per day with  Teams....ZambiaEasternSindaChiwuyuNameStructures CountRiskTarget StatusStatus setting by...Selected JurisdictionsCWY_2517880TargetedAuto-selection1CWY_3241--Not Targeted0CWY_14336--Not Targeted0CWY_2297--Not Targeted0CWY_829--Not Targeted0CWY_15483--Not Targeted0CWY_715--Not Targeted0CWY_18266--Not Targeted0CWY_1126--Not Targeted0CWY_12269--Not Targeted0CWY_19304--Not Targeted0CWY_13251--Not Targeted0CWY_19015--Not Targeted0CWY_24185--Not Targeted0CWY_4299--Not Targeted0CWY_1017--Not Targeted0CWY_613--Not Targeted0CWY_26139--Not Targeted0CWY_16320--Not Targeted0CWY_11266--Not Targeted0CWY_17167--Not Targeted0CWY_23290--Not Targeted0CWY_5168--Not Targeted0save draftSave & Activate"`;

exports[`e2e test auto-selection flow from slider auto-selection to refine jurisdictions: should have single node(parent node) 1`] = `"NameStructures Count"`;

exports[`e2e test auto-selection flow from slider auto-selection to refine jurisdictions: should have single node(parent node) 2`] = `"Zambia178"`;

exports[`e2e test auto-selection flow from slider auto-selection to refine jurisdictions: should have single node(parent node) 3`] = `"Total178"`;

exports[`e2e test auto-selection flow from slider auto-selection to refine jurisdictions: should not have btn-link or onClick handler 1`] = `
<span
  key="f6baee38-43af-42a9-941b-a15a5499ce35-span"
>
  CWY_25
</span>
`;

exports[`e2e test auto-selection flow from slider auto-selection to refine jurisdictions: structure summary step snapshot 1`] = `"Planning toolA2-Lusaka Akros Test Focus 2A2-Lusaka Akros Test Focus 2Auto target jurisdictions by riskRefine selected jurisdictionsAssignment WrapperResource Estimate for Zambia0 daysat a rate of   structures per team per day with  Teams....NameStructures CountZambia178Total178Continue to next step"`;

exports[`e2e test auto-selection flow when there are no selected jurisdictions(without metaData): full page rendered text 1`] = `"Planning toolA2-Lusaka Akros Test Focus 2A2-Lusaka Akros Test Focus 2Auto target jurisdictions by riskRefine selected jurisdictionsAssignment WrapperZambia:  There are no descendant jurisdictions that are targeted.....NameStructures CountRiskTarget StatusStatus setting by...Selected JurisdictionsZambia575280--0save draftSave & Activate"`;

exports[`e2e test auto-selection flow when there are no selected jurisdictions(without metaData): general page snapshot 1`] = `"Planning toolA2-Lusaka Akros Test Focus 2A2-Lusaka Akros Test Focus 2Auto target jurisdictions by riskRefine selected jurisdictionsAdjust slider to auto-target jurisdictionsRisk  0%00100NUMBER OF STRUCTURES IN SELECTED JURISDICTIONS0Continue to next step"`;

exports[`e2e test auto-selection flow when there are no selected jurisdictions(without metaData): structure summary step snapshot 1`] = `"Planning toolA2-Lusaka Akros Test Focus 2A2-Lusaka Akros Test Focus 2Auto target jurisdictions by riskRefine selected jurisdictionsAssignment WrapperZambia:  There are no descendant jurisdictions that are targeted.....NameStructures CountThere are no descendant jurisdictions that are targeted.Continue to next step"`;

exports[`e2e test auto-selection flow when there are no selected jurisdictions(without metaData): will not have any nodes 1`] = `"NameStructures Count"`;
